x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,704 @@
+                return;
+            }
+            final AuthenticationSuccessResponse response = authenticationResponse.toSuccessResponse();
+            validateState(expectedState, response.getState());
+            validateResponseType(response);
+            if (rpConfig.getResponseType().impliesCodeFlow()) {
+                final AuthorizationCode code = response.getAuthorizationCode();
+                exchangeCodeForToken(code, ActionListener.wrap(tokens -> {
+                    final AccessToken accessToken = tokens.v1();
+                    final JWT idToken = tokens.v2();
+                    validateAccessToken(accessToken, idToken);
+                    getUserClaims(accessToken, idToken, expectedNonce, true, listener);
+                }, listener::onFailure));
+            } else {
+                final JWT idToken = response.getIDToken();
+                final AccessToken accessToken = response.getAccessToken();
+                validateAccessToken(accessToken, idToken);
+                getUserClaims(accessToken, idToken, expectedNonce, true, listener);
+            }
+        } catch (ElasticsearchSecurityException e) {
+            // Don't wrap in a new ElasticsearchSecurityException
+            listener.onFailure(e);
+        } catch (Exception e) {
+            listener.onFailure(new ElasticsearchSecurityException("Failed to consume the OpenID connect response. ", e));
+        }
+    }
+
+    /**
+     * Collects all the user claims we can get for the authenticated user. This happens in two steps:
+     * <ul>
+     * <li>First we attempt to validate the Id Token we have received and get any claims it contains</li>
+     * <li>If we have received an Access Token and the UserInfo endpoint is configured, we also attempt to get the user info response
+     * from there and parse the returned claims,
+     * see {@link OpenIdConnectAuthenticator#getAndCombineUserInfoClaims(AccessToken, JWTClaimsSet, ActionListener)}</li>
+     * </ul>
+     *
+     * @param accessToken    The {@link AccessToken} that the OP has issued for this user
+     * @param idToken        The {@link JWT} Id Token that the OP has issued for this user
+     * @param expectedNonce  The nonce value we sent in the authentication request and should be contained in the Id Token
+     * @param claimsListener The listener to notify with the resolved {@link JWTClaimsSet}
+     */
+    private void getUserClaims(@Nullable AccessToken accessToken, JWT idToken, Nonce expectedNonce, boolean shouldRetry,
+                               ActionListener<JWTClaimsSet> claimsListener) {
+        try {
+            JWTClaimsSet verifiedIdTokenClaims = idTokenValidator.validate(idToken, expectedNonce).toJWTClaimsSet();
+            if (logger.isTraceEnabled()) {
+                logger.trace("Received and validated the Id Token for the user: [{}]", verifiedIdTokenClaims);
+            }
+            if (accessToken != null && opConfig.getUserinfoEndpoint() != null) {
```

Comment:
if `accessToken != null` && `opConfig.getUserinfoEndpoint() == null` we do nothing. I think at least logging should be done. For example if the URL is wrong, `getAndCombineUserInfoClaims` would throw `ElasticsearchSecurityException` .

Reply:
`accessToken != null && opConfig.getUserinfoEndpoint() == null` is a valid case where the OP returned an access token but we haven't configured a userInfoEndpoint in our settings because we are not interested in hitting it ( all necessary claims are in the ID token ). The same goes for `accessToken == null && opConfig.getUserinfoEndpoint() != null`: We have configured an userinfo endpoint but the OP didn't return an access token (maybe we misconfigured our response type to `id_token` instead of `token id_token`, so we can't make a request to the user info endpoint. Logging the latter wouldn't hurt, you're right :)

> For example if the URL is wrong, getAndCombineUserInfoClaims would throw ElasticsearchSecurityException .

If the URL is invalid, we would throw on building the realm config. If it is valid but wrong, we would catch that in getAndCombineUserInfoClaims and call `onFailure(new ElasticsearchSecurityException("Failed to get claims from the Userinfo Endpoint.",`